### ISSUE TITLE
Issue with Sky.cpp when building PlatoSim from source

### ISSUE THREAD

**Issue by jmccormac01** [IvS-KULeuven/PlatoSim3#1012]
Hi, 
I am working on implementing [this feature request](https://github.com/IvS-KULeuven/PlatoSim3/issues/955). I started back in July before traveling and holiday. I'm working on this on [a branch in my fork](https://github.com/jmccormac01/PlatoSim3/commits/platonium_new_l1_pipeline/) for now.

I noticed today that the upstream/develop has 22 commits so I merged them into my feature branch. After doing so, installing from source now fails with the following error:

```sh
[ 75%] Building CXX object CMakeFiles/platosimLibrary.dir/source/Random.cpp.o
At global scope:
cc1plus: warning: unrecognized command line option '-Wno-deprecated-register'
[ 77%] Building CXX object CMakeFiles/platosimLibrary.dir/source/Simulation.cpp.o
[ 79%] Building CXX object CMakeFiles/platosimLibrary.dir/source/Sky.cpp.o
[ 81%] Building CXX object CMakeFiles/platosimLibrary.dir/source/SkyCoordinates.cpp.o
[ 83%] Building CXX object CMakeFiles/platosimLibrary.dir/source/StringUtilities.cpp.o
[ 85%] Building CXX object CMakeFiles/platosimLibrary.dir/source/Telescope.cpp.o
[ 87%] Building CXX object CMakeFiles/platosimLibrary.dir/source/TemperatureFromFile.cpp.o
/ecosystem/PlatoSim3/source/Sky.cpp: In member function 'std::tuple<double, double, double> Sky::getInfoOfStarWithID(unsigned int)':
/ecosystem/PlatoSim3/source/Sky.cpp:794:28: error: converting to 'std::tuple<double, double, double>' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {int, int, int}; <template-parameter-2-2> = void; _Elements = {double, double, double}]'
             return {0, 0, 0};
                            ^
At global scope:
cc1plus: warning: unrecognized command line option '-Wno-deprecated-register'
make[2]: *** [CMakeFiles/platosimLibrary.dir/source/Sky.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/platosimLibrary.dir/build.make:902: recipe for target 'CMakeFiles/platosimLibrary.dir/source/Sky.cpp.o' failed
/ecosystem/PlatoSim3/source/TemperatureFromFile.cpp: In member function 'virtual double TemperatureFromFile::getNextTemperature(double)':
/ecosystem/PlatoSim3/source/TemperatureFromFile.cpp:142:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         if (timeIndex >= timeFromFile.size())
                       ^
At global scope:
cc1plus: warning: unrecognized command line option '-Wno-deprecated-register'
make[1]: *** [CMakeFiles/platosimLibrary.dir/all] Error 2
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/platosimLibrary.dir/all' failed
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
The command '/bin/bash --login -c ./install.sh' returned a non-zero code: 2
``` 

The issue seems to come from `Sky.cpp` that was updated 3w ago while I was away [here](https://github.com/IvS-KULeuven/PlatoSim3/commits/develop/source/Sky.cpp) 

I wanted to point this out. I can roll back the merge to continue testing, but at the minute PS is not building from source for me any longer.

**Comment by nicholasjannsen** [comment]
Hi James, thanks for this update. It could be a mistake in `Sky.cpp` while we rolled back some changes. I cannot access your branch, did you push it to upstream? Then I can see if I can reproduce the result (and hopefully fix it).

FYI I'm on the commit: 6759e0372b0a1e99959ec682ba02f482a404b197
Building PlatoSim develop works fine for me here (I think this is to commits further than your branch).

**Comment by JorisDeRidder** [comment]
Can you post which compiler and which version you use to compile the PlatoSim source code?

**Comment by jmccormac01** [comment]
This is the first bit of output from `install.sh`. I'm building in a docker container and can share the full output if needed. 

```
compiler 'gcc' is GNU gcc-5.4.0
compiler 'gfortran' is GNU gfortran-5.4.0
compiler 'g++' is GNU g++-5.4.0
checking for config ./config/site-specific/host-316a556e6d94... no
checking build mode... debug
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking if unsupported combinations of configure options are allowed... no
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for off_t... yes
checking for size_t... yes
checking for ssize_t... yes
checking for ptrdiff_t... yes
```

**Comment by nicholasjannsen** [comment]
Hi James, I checked and I use gcc version 13.1.0. Would it be possible for you to try to update your gcc compilers?

**Comment by jmccormac01** [comment]
Oh, that's quite the difference in versions. I'll give that a go and let you know. Thanks.

**Comment by jmccormac01** [comment]
@nicholasjannsen I've tried updating my docker base OS from (a very old version of) ubuntu:16.04 to get more modern compilers. I've tried using 24.04, 22.04 and 20.04 each of which lead to issues with mpi4py. 

```Note: This error originates from the build backend, and is likely not a problem with poetry but with mpi4py (3.1.6) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "mpi4py (==3.1.6)"'.```

Can I check what base OS/version you develop on and what python version you recommend. I can try and match my containers to your env as best possible to reduce potential sources of problems.

**Comment by nicholasjannsen** [comment]
Hi James, sorry for the delay.. I also use 20.04 (haven't dared to update my system the past year).

I recommend to use Python 3.9. I specifically create a new conda environment with:
```
conda create -n blabla python=3.9
``` 
and then use Poetry to install of all Python dependencies:
```
poetry install --with platonium --with pipeline
```
I just tested the above command on my system and everything installs perfectly fine:
```
(test) nicholas@messier:~/software/PlatoSim3$ poetry install --with platonium --with pipeline
Updating dependencies
Resolving dependencies... (63.4s)

Package operations: 168 installs, 1 update, 0 removals

  • Installing attrs (24.2.0)
  • Installing rpds-py (0.20.0)
  • Installing pycparser (2.22)
  • Installing referencing (0.35.1)
  • Installing certifi (2024.8.30)
  • Installing cffi (1.17.1)
  • Installing charset-normalizer (3.3.2)
  • Installing idna (3.8)
  • Installing jsonschema-specifications (2023.12.1)
  • Installing numpy (1.23.5)
  • Installing platformdirs (4.3.2)
  • Installing six (1.16.0)
  • Installing traitlets (5.9.0)
  • Installing urllib3 (2.2.3)
  • Installing asttokens (2.4.1)
  • Installing backports-tarfile (1.2.0)
  • Installing cryptography (43.0.1)
  • Installing entrypoints (0.4)
  • Installing executing (2.1.0)
  • Installing fastjsonschema (2.20.0)
  • Installing jeepney (0.8.0)
  • Installing jsonschema (4.23.0)
  • Installing jupyter-core (5.7.2)
  • Installing locket (1.0.0)
  • Installing more-itertools (10.5.0)
  • Installing nest-asyncio (1.6.0)
  • Installing packaging (24.1)
  • Installing parso (0.8.4)
  • Installing ptyprocess (0.7.0)
  • Installing pure-eval (0.2.3)
  • Installing pyerfa (2.0.1.4)
  • Installing pyjwt (2.9.0)
  • Installing python-dateutil (2.9.0.post0)
  • Installing pyyaml (6.0.2)
  • Installing pyzmq (26.2.0)
  • Installing requests (2.32.3)
  • Installing toolz (0.12.1)
  • Installing tornado (6.1)
  • Installing wcwidth (0.2.13)
  • Installing zipp (3.20.1)
  • Installing asciitree (0.3.3)
  • Installing astropy (5.3.4)
  • Installing click (8.1.7)
  • Installing cloudpickle (3.0.0)
  • Installing contourpy (1.3.0)
  • Installing cycler (0.12.1)
  • Installing decorator (5.1.1)
  • Installing exceptiongroup (1.2.2)
  • Installing fasteners (0.19)
  • Installing fonttools (4.53.1)
  • Installing fsspec (2024.9.0)
  • Installing importlib-metadata (8.5.0)
  • Installing importlib-resources (6.4.5)
  • Installing jaraco-classes (3.4.0)
  • Installing jaraco-context (6.0.1)
  • Installing jaraco-functools (4.0.2)
  • Installing jedi (0.19.1)
  • Installing jupyter-client (7.3.2)
  • Installing kiwisolver (1.4.7)
  • Installing ligo-segments (1.4.0)
  • Installing markupsafe (2.1.5)
  • Installing matplotlib-inline (0.1.7)
  • Installing nbformat (5.10.4)
  • Installing numcodecs (0.12.1)
  • Installing partd (1.4.2)
  • Installing pexpect (4.9.0)
  • Installing pillow (10.4.0)
  • Installing prompt-toolkit (3.0.47)
  • Installing pygments (2.18.0)
  • Installing pyopenssl (24.2.1)
  • Installing pyparsing (3.1.4)
  • Installing safe-netrc (1.0.1)
  • Installing scitokens (1.8.1)
  • Installing secretstorage (3.3.3)
  • Installing soupsieve (2.6)
  • Installing stack-data (0.6.3)
  • Installing typing-extensions (4.12.2)
  • Installing webencodings (0.5.1)
  • Installing argon2-cffi-bindings (21.2.0)
  • Installing astropy-healpix (1.0.3)
  • Installing beautifulsoup4 (4.12.3)
  • Installing bleach (6.1.0)
  • Installing comm (0.2.2)
  • Installing dask (2024.8.0)
  • Installing debugpy (1.8.5): Downloading... 50%
  • Installing debugpy (1.8.5)
  • Installing defusedxml (0.7.1)
  • Installing future (1.0.0)
  • Installing html5lib (1.1)
  • Installing igwn-auth-utils (1.1.1)
  • Installing ipython (8.18.1)
  • Installing jinja2 (3.1.4)
  • Installing jupyterlab-pygments (0.3.0)
  • Installing keyring (25.3.0)
  • Installing llvmlite (0.39.1)
  • Installing lscsoft-glue (4.0.0)
  • Installing matplotlib (3.9.2)
  • Installing mistune (3.0.2)
  • Installing nbclient (0.10.0)
  • Installing pandocfilters (1.5.1)
  • Installing psutil (6.0.0)
  • Installing pyqtgraph (0.13.7)
  • Installing pytz (2024.2)
  • Installing pyvo (1.5.2)
  • Installing scipy (1.13.1)
  • Updating setuptools (73.0.1 -> 74.1.2)
  • Installing tinycss2 (1.3.0)
  • Installing tqdm (4.66.5)
  • Installing zarr (2.18.2)
  • Installing argon2-cffi (23.1.0)
  • Installing asteval (1.0.2)
  • Installing astroplan (0.10.1)
  • Installing astroquery (0.4.7)
  • Installing batman-package (2.4.9)
  • Installing bidict (0.23.1)
  • Installing cmlib (1.1.3)
  • Installing configparser (7.1.0)
  • Installing dill (0.3.8)
  • Installing h5py (3.11.0)
  • Installing healpy (1.17.3)
  • Installing imageio (2.35.1)
  • Installing ipykernel (6.29.5)
  • Installing ipython-genutils (0.2.0)
  • Installing jupyterlab-widgets (3.0.13)
  • Installing lalsuite (7.23)
  • Installing ligo-gracedb (2.12.0)
  • Installing nbconvert (7.16.4)
  • Installing networkx (3.2.1)
  • Installing numba (0.56.4)
  • Installing pandas (1.5.3)
  • Installing patsy (0.5.6)
  • Installing pgcolorbar (1.1.3)
  • Installing prometheus-client (0.20.0)
  • Installing ptemcee (1.0.0)
  • Installing pyqt5-sip (12.15.0)
  • Installing python-ligo-lw (1.8.3)
  • Installing pywavelets (1.6.0)
  • Installing quantities (0.16.0)
  • Installing reproject (0.13.0)
  • Installing semantic-version (2.10.0)
  • Installing send2trash (1.8.3)
  • Installing tenacity (9.0.0)
  • Installing terminado (0.18.1)
  • Installing tifffile (2024.8.30)
  • Installing uncertainties (3.2.2)
  • Installing widgetsnbextension (4.0.13)
  • Installing zope-interface (7.0.3)
  • Installing argos (0.4.4)
  • Installing colorama (0.4.6)
  • Installing cython (0.29.37)
  • Installing datetime (4.9)
  • Installing descartes (1.1.0)
  • Installing ipywidgets (8.1.5)
  • Installing joblib (1.4.2)
  • Installing ldtk (1.8.4)
  • Installing ligo-skymap (0.6.1)
  • Installing lmfit (1.3.2)
  • Installing mpi4py (3.1.6)
  • Installing natsort (8.4.0)
  • Installing notebook (6.4.6)
  • Installing prettytable (3.11.0)
  • Installing pyaml (21.10.1)
  • Installing pyarrow (13.0.0)
  • Installing pyastronomy (0.17.1)
  • Installing pyqt5 (5.15.2)
  • Installing scikit-image (0.19.3)
  • Installing shapely (1.8.5.post1)
  • Installing statsmodels (0.13.5)
  • Installing transitleastsquares (1.32)
  • Installing wotan (1.10)

Writing lock file

Installing the current project: platosim (3.6.1)
```
 If you use Poetry, then you can use the `poetry.lock` file to install the packages above. Strangely enough I do not see the install in the above execution, however, looking into the `poetry.lock` file it is indeed installed with the version `3.1.6`. GitHub do not support `poetry.lock` files so below is a txt file you simply need to rename to the original file and then place into the base of PlatoSim3.
[poetry.lock.txt](https://github.com/user-attachments/files/16985895/poetry.lock.txt)

**Comment by jmccormac01** [comment]
It turns out I was having endless issues with poetry itself, likely due to the way I was installing it inside the docker containers. I've spent all afternoon reading about it and I'm now starting my dockerfile over from scratch. I've got a fresh python3.9 install on ubuntu 20.04 and I've got poetry installing the ```pyproject.toml``` file fine now. I'll keep going with the rest of the PlatoSim build, fingers crossed.
